lib/instrumentation/modules/@aws-sdk/client-s3.js

```diff
@@ -61,14 +61,15 @@ function s3MiddlewareFactory (client, agent) {
   return [
     {
       middleware: (next, context) => async (args) => {
-        const input = args.input
-        const bucket = input && input.Bucket
-        const resource = resourceFromBucket(bucket)
+        // Ensure there is a span from the wrapped `client.send()`.
         const span = agent._instrumentation.currSpan()
-
-        if (!span) {
+        if (!span || !(span.type === TYPE && span.subtype === SUBTYPE)) {
```

Comment:
Reviewer note: This is the fix for the first surprise. We do a slightly weak check that the current span is one likely created by our `client.send()` wrapper (in smithy-client.js).